ecu/stats.go

```diff
@@ -32,133 +32,128 @@ import (
 	log "github.com/golang/glog"
 )
 
-// publisher implements an interface for a statistics publisher.
-type publisher interface {
-	update(s *stats)
+// Publisher is an interface for a statistics publisher. Publishers are 
+// notified of updates to the statistics with the Update method.
+type Publisher interface {
+	Update(s *Stats)
 }
 
-// stats contains the statistics collected from the Seesaw Engine.
-type stats struct {
-	lock sync.RWMutex
-
-	lastUpdate  time.Time
-	lastSuccess time.Time
-
-	seesaw.ClusterStatus
-	seesaw.ConfigStatus
-	seesaw.HAStatus
-	neighbors []*quagga.Neighbor
-	vlans     []*seesaw.VLAN
-	vservers  map[string]*seesaw.Vserver
+// Stats contains the statistics collected from the Seesaw Engine.
+type Stats struct {
+	LastUpdate  time.Time
+	LastSuccess time.Time
+
+	ClusterStatus seesaw.ClusterStatus
+	ConfigStatus  seesaw.ConfigStatus
+	HAStatus      seesaw.HAStatus
+	Neighbors     []*quagga.Neighbor
+	VLANs         []*seesaw.VLAN
+	Vservers      map[string]*seesaw.Vserver
 }
 
 // ecuStats contains that data needed to the ECU stats collector.
 type ecuStats struct {
 	ecu        *ECU
-	publishers []publisher
-	stats      *stats
+
+	publishersMu sync.RWMutex
+	publishers []Publisher
+
+	lastUpdateMu sync.RWMutex
+	lastUpdate   time.Time
```

Comment:
what's the usage of this lastUpdate?

Reply:
To enable distinguishing a problem with gathering stats versus a problem with ECU not running correctly, in a monitoring system, by comparing the progression of timestamps. e.g. LastUpdate increasing but LastSuccess not increasing. I've changed it, hopefully it's more sensible now.